crypto/evp/evp_rand.c

```diff
@@ -0,0 +1,532 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "internal/cryptlib.h"
+#include <openssl/engine.h>
+#include <openssl/evp.h>
+#include <openssl/x509v3.h>
+#include <openssl/rand.h>
+#include <openssl/core.h>
+#include <openssl/core_names.h>
+#include <openssl/crypto.h>
+#include "crypto/asn1.h"
+#include "crypto/evp.h"
+#include "internal/numbers.h"
+#include "internal/provider.h"
+#include "evp_local.h"
+
+static int evp_rand_up_ref(void *vrand)
+{
+    EVP_RAND *rand = (EVP_RAND *)vrand;
+    int ref = 0;
+
+    if (rand != NULL)
+        return CRYPTO_UP_REF(&rand->refcnt, &ref, rand->lock);
+    return 1;
+}
+
+static void evp_rand_free(void *vrand){
+    EVP_RAND *rand = (EVP_RAND *)vrand;
+    int ref = 0;
+
+    if (rand != NULL) {
+        CRYPTO_DOWN_REF(&rand->refcnt, &ref, rand->lock);
+        if (ref <= 0) {
+            ossl_provider_free(rand->prov);
+            CRYPTO_THREAD_lock_free(rand->lock);
+            OPENSSL_free(rand);
+        }
+    }
+}
+
+static void *evp_rand_new(void)
+{
+    EVP_RAND *rand = NULL;
+
+    if ((rand = OPENSSL_zalloc(sizeof(*rand))) == NULL
+        || (rand->lock = CRYPTO_THREAD_lock_new()) == NULL) {
+        evp_rand_free(rand);
+        return NULL;
+    }
+    rand->refcnt = 1;
+    return rand;
+}
+
+/* Enable locking of the underlying DRBG/RAND if available */
+int EVP_RAND_CTX_enable_locking(EVP_RAND_CTX *rand)
+{
+    if (rand->meth->enable_prov_locking != NULL)
+        return rand->meth->enable_prov_locking(rand->data);
+    return 1;
```

Comment:
is "1" the okay, or is it an error?  if there's now provider locking, seems like this should return an error.  here, and below.

Reply:
It is okay: a RNG that cannot enable locking is saying that it doesn't need it.
I'll document this.

Reply:
After more thought, I've changed this to an error return.